src/Elmish/Hooks/Type.js

```diff
@@ -0,0 +1,3 @@
+const uuidV5 = require('uuid/v5')
+
+exports.genStableUUID = () => uuidV5(new Error().stack, '31877c6f-998d-44e6-99e6-3cd31a643f1d')
```

Comment:
So here lies the issue. I don't think the _full_ stack is a good fingerprint, because it might be different if the component is used from different places. Consider this:

```
view = if foo then bar () else qux ()
  where
    bar () = fragment [s (), div "" "Some stuff"]
    qux () = fragment [s (), div "" "Other stuff"]

    s () = useState ...
```

Here, I intend `s` to be the same component, but it's accessed via different stacks - via `bar` and via `qux`, depending on `foo`, - which means that React will consider `bar -> s` to be one component and `qux -> s` to be a completely different component. And this means that changing `foo` will cause `s`'s state to get reset.

Reply:
Something similar happens in `Hub.Frame` and other similar places.

Reply:
So the fingerprint should be literally just the `useState` call site, nothing above it.

Reply:
Oh I see. In this case, I’m not sure it will be possible for this library to know when the consumer wants it to be the same component. In my examples, there was usually only one line different in the stacks. And given your example, I can think of other ways to represent that where it wouldn’t be obvious that the intent was they are the same component.

What if we made it so that we generate a name by default, but then allow a keyed approach, like React does with components?

```purs
view = withHooks do
  … <- useState ""
  pure if foo then bar () else qux ()
  where
    bar () = fragment [s (), div "" "Some stuff"]
    qux () = fragment [s (), div "" "Other stuff"]

    s () = useState' (Key "MyHook") ...
```

Reply:
> So the fingerprint should be literally just the useState call site, nothing above it.

Oh, I see. I think that makes sense. I can try that.

Reply:
For future reference, test code for Fyodor’s example:

```purs
-- examples/src/Main.purs
view :: Unit -> Dispatch Void -> ReactElement
view _ _ = withHooks do
  foo /\ setFoo <- useState true
  pure if foo then bar foo setFoo else qux foo setFoo
  where
    bar foo setFoo = H.fragment [s unit, H.div_ "" { onClick: setFoo $ not foo } "Some stuff"]
    qux foo setFoo = H.fragment [s unit, H.div_ "" { onClick: setFoo $ not foo } "Other stuff"]

    s _ = withHooks do
      value /\ setValue <- useState ""
      pure $ H.input_ "form-control" { value, onChange: setValue <?| eventTargetValue }
```

Reply:
### Before Suggested Change

![Dec-23-2021 01-40-49](https://user-images.githubusercontent.com/23478119/147199492-a818a1e0-66c2-45ed-b802-2762ab304527.gif)

### After Suggested Change

![Dec-23-2021 01-39-19](https://user-images.githubusercontent.com/23478119/147199506-bb2ce685-69cd-45ca-bda9-8b210204e8ec.gif)